common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -341,7 +373,38 @@ private static int collationNameToId(String collationName) throws SparkException
         }
       }
 
+      /**
+       * The catalog to which the `CollationSpec` belongs
+       */
+      protected abstract String catalog();
+
+      /**
+       * The schema to which the `CollationSpec` belongs
+       */
+      protected abstract String schema();
+
       protected abstract Collation buildCollation();
+
+      static List<Collation> listCollations(String catalog, String schema, String pattern) {
+        return Stream.concat(
+          CollationSpecUTF8.listCollations(catalog, schema, pattern).stream(),
+          CollationSpecICU.listCollations(catalog, schema, pattern).stream()).toList();
+      }
+
+      static List<String> filterPattern(List<String> names, String pattern) {
```

Comment:
Could we make use of StringUtils in the actual command to filterPattern, as this feels like code duplication.

Reply:
There are two difficulties here,
- If we use `StringUtils.filterPattern` in the command `ShowCollationsCommand`, the prerequisite is that we can first enumerate all `collations`, and as commented by https://github.com/apache/spark/pull/47364#discussion_r1686142045, this is too expensive
- Class `StringUtils` is currently in module `spark catalyst`, and we want to use method `StringUtils.filterPattern` in module `spark unsafe`, but obviously we cannot directly use it. (the module `spark catalyst` depends on `spark unsafe`, not `spark unsafe` depends on `spark catalyst`)

Reply:
Could we just extract collation names? As in example with functions, we treat built-in functions separately by just listing their names from FunctionRegistry. It was expensive to list all Collation objects as they contain Collators which are externally used objects. For the other part, if we move the filtering process to `SessionCatalog`, similar to functions we should be good with the package problem.

Reply:
I think it's okay, I'll give it a try.